python/remote_api/src/lynx/kite.py

```diff
@@ -854,6 +874,55 @@ def set_executors(self, count):
     return self._send('/remote/setExecutors', {'count': count})
 
 
+def get_free_ports(n):
+  servers = [socketserver.TCPServer(('localhost', 0), None) for _ in range(n)]
+  for s in servers:
+    s.__enter__()
+  ports = [s.server_address[1] for s in servers]
+  for s in servers:
+    s.__exit__()
+  return ports
+
+
+class ManagedLynxKite:
+  def __init__(self, spark, **kwargs):
+    self.assert_jar_is_loaded(spark)
+    self.spark = spark
+    self.storage = TemporaryDirectory()
+    tmp = self.storage.name
+    kite_http_port, kite_https_port, sphynx_port = get_free_ports(3)
+    self.set_env(
+        KITE_META_DIR=f'{tmp}/meta',
+        KITE_DATA_DIR=f'file:{tmp}/data',
+        KITE_HTTP_PORT=kite_http_port,
+        KITE_HTTPS_PORT=kite_https_port,
+        SPHYNX_HOST='localhost',
+        SPHYNX_PORT=sphynx_port,
+        ORDERED_SPHYNX_DATA_DIR=f'{tmp}/sphynx/ordered',
+        UNORDERED_SPHYNX_DATA_DIR=f'{tmp}/sphynx/unordered',
+        **kwargs)
+    self.address = f'http://localhost:{kite_http_port}'
+
+  def assert_jar_is_loaded(self, spark):
+    assert (
+        not isinstance(spark._jvm.com.lynxanalytics.biggraph.LynxKite, type(spark._jvm.com.nonexistent.package))), \
+        'The LynxKite jar has not been loaded in this SparkSession.'
+
+  def set_env(self, **kwargs):
+    jvm = self.spark._jvm
+    cfg = jvm.java.util.ArrayList()
+    for k, v in kwargs.items():
+      cfg.append(jvm.Tuple2(k, str(v)))
+    cfg = jvm.scala.collection.JavaConversions.asScalaBuffer(cfg)
+    jvm.com.lynxanalytics.biggraph.Environment.set(cfg)
+
+  def start(self):
+    self.spark._jvm.com.lynxanalytics.biggraph.LynxKite.start()
```

Comment:
When does this function return?

Reply:
> When does this function return?

It appears to return when LynxKite is ready to receive requests. I've added a docstring to make this official.

I have to admit I'm not sure what ensures this. In `LynxKite.scala` we call `play.core.server.ProdServerStart.start()`. My guess is it only returns when everything is ready.

The reason I think this works correctly is because the tests pass reliably.